Project Plugins: Add Plugin Service

luispadron opened the pull request:
### Short description 📝

This is another PR following up on work implemented in: #2095, #2103, #2107

#### New modules

In this PR we add two new modules: `TuistPlugin` and `TuistPluginTests`. 

The `TuistPlugin` module is intended to hold code related to tuist plugin feature. This includes the newly added `PluginService`. 

#### PluginService

With this PR we also add `PluginServicing` which is an interface for fetching and loading plugins. This service intends to allow developers to load `Plugin` manifests. In order to first load a plugin manifest we must determine where the manifest is located.
- For cases where the plugin is `local` we simply search in the specified path.
- For cases where the plugin is a "remote" plugin, i.e. a git plugin we use the `GitService` implemented in #2107. We fetch the plugin at the remote and store it into the tuist cache directory under the `/Plugins` directory. The fetched repository is then fingerprinted (using the remote URL and "git id"). With the plugin fetched we simply load the manifest using the loader.

#### Note

This PR does **not** add any actual functionality for loading a `Plugin` it simply adds the `PluginService` and any required changes to get this compiling. There will be a follow up PR for the logic required to load plugin manifests.

luispadron commented on Sources/TuistSupport/Constants.swift:
```diff
@@ -1,6 +1,6 @@
 import Foundation
 
-public struct Constants {
+public enum Constants {
```
Changed these from struct to enums since they just hold constants.

pepibumur commented on Sources/TuistPlugin/PluginService.swift:
```diff
@@ -0,0 +1,89 @@
+/// A default implementation of `PluginServicing` which loads `Plugins` using
+/// the `Config` description without it needing to be loaded. The  plugins are first fetched/copied into
+/// a cache and the loaded `Plugins` model is returned.
+public final class PluginService: PluginServicing {
```
To be consistent with the other files in the codebase and prevent back-and-forths between files, I'd recommend keeping `PluginService`, `PluginServicing`, and `PluginServiceError` in the same file.

pepibumur commented on Sources/TuistPlugin/PluginService.swift:
```diff
@@ -0,0 +1,89 @@
+
+    /// fetches the git plugins from the remote server and caches them in
+    /// the Tuist cache with a unique fingerprint
+    private func fetchGitPlugin(at url: String, with gitId: String) throws -> AbsolutePath {
```
Nitpick: Should we rename `gitId` to `gitReference` to be consistent with [Git's terminology](https://git-scm.com/book/en/v2/Git-Internals-Git-References)

pepibumur commented on Sources/TuistPlugin/PluginService.swift:
```diff
@@ -0,0 +1,89 @@
+    /// the Tuist cache with a unique fingerprint
+    private func fetchGitPlugin(at url: String, with gitId: String) throws -> AbsolutePath {
+        let fingerprint = "\(url)-\(gitId)".md5
+        let pluginDirectory = Environment.shared.cacheDirectory
```
Nitpick: What about adding a new method to `Environment` to return the plugins directory?

pepibumur commented on Sources/TuistPlugin/PluginService.swift:
```diff
@@ -0,0 +1,89 @@
+                case let .local(path):
+                    logger.debug("Fetching \(plugin.description) at: \(path)")
+                    return AbsolutePath(path)
+                case let .gitWithBranch(url, id),
```
Support for branches requires further work because we might have a local copy pointing to a commit X, but in remote there's a more recent commit Y that we might want to fetch. For this first implementation, I'd only support tags and shas. Once we see how users use the feature, if they ask for branches support, we can invest some efforts into enabling it. What do you think?

pepibumur reviewed (approved):
Solid work @luispadron 👏

pepibumur commented:
@all-contributors add @luispadron for code

fortmarek commented on Tests/TuistPluginTests/PluginServiceTests.swift:
```diff
@@ -0,0 +1,134 @@
+
+    override func setUp() {
+        super.setUp()
+        do {
```
I think it's more transparent to use `setUpWithError() throws` function and do not use the explicit `XCTFail` here

fortmarek commented on Tests/TuistPluginTests/PluginServiceTests.swift:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    override func tearDown() {
+        super.tearDown()
```
nullify the variables of this test class, please 🙂

fortmarek commented on Tests/TuistPluginTests/PluginServiceTests.swift:
```diff
@@ -0,0 +1,134 @@
+        super.tearDown()
+    }
+
+    func test_loadPlugins_atPath_WHEN_localHelpers() throws {
```
In most test cases we follow the `// Given // When // Then` convention, so I'd add that to the test cases in this file.

fortmarek commented on Tests/TuistPluginTests/PluginServiceTests.swift:
```diff
@@ -0,0 +1,134 @@
+@testable import TuistPlugin
+
+final class PluginServiceTests: TuistTestCase {
+    private var mockModelLoader: MockGeneratorModelLoader!
```
I'd drop the `mock` prefix as it's redundant here IMHO -> it's a unit test, so it's safe to assume that all the implementations will be mock ones

fortmarek commented on Tests/TuistPluginTests/PluginServiceTests.swift:
```diff
@@ -0,0 +1,134 @@
+        XCTAssertEqual(plugins, expectedPlugins)
+    }
+
+    private func mockConfig(plugins: [PluginLocation]) -> Config {
```
Is there a reason you are not using `Config.test()` extension function defined in `Config+TestData.swift` ?

luispadron replied:
Ah right we I should totally use that one instead!

fortmarek reviewed (commented):
Sorry for the longer delay of my review, the changes seem solid, only a few nitpicks 🎉 2021 will be the year of Tuist plugins!

luispadron commented on Sources/TuistPlugin/PluginService.swift:
```diff
@@ -0,0 +1,89 @@
+    private func fetchGitPlugin(at url: String, with gitId: String) throws -> AbsolutePath {
+        let fingerprint = "\(url)-\(gitId)".md5
+        let pluginDirectory = Environment.shared.cacheDirectory
+            .appending(RelativePath(Constants.PluginDirectory.name))
```
Yep! I keep forgetting about this function

luispadron commented:
@pepibumur @fortmarek  Thank you both for the review! I've addresses most of your comments in the subsequent commit. 

As for:

> Support for branches requires further work because we might have a local copy pointing to a commit X, but in remote there's a more recent commit Y that we might want to fetch. For this first implementation, I'd only support tags and shas. Once we see how users use the feature, if they ask for branches support, we can invest some efforts into enabling it. What do you think?

@pepibumur I think I see I understand what you mean here, are you saying that for branch references to plugins we will need to ensure that not the cached version of the plugin is up-to-date with the upstream git branch? If so I think I can add support for this now. Do you think it would be better to just ship this plugins feature with all 3 references supported or add it in later?

Thinking about what is required: 

- We could (on top of checking if we have the repo in the cache) also check that the repo points to the correct location.

To do this we can either:

- Pull every time (since this would be expected if the user is using `branch` reference, this means they want the latest changes of that given branch)

or:

- Perform some fancy git commands to determine if we need to pull or not and if we do perform that pull.

What do you all think?

pepibumur commented:
> Do you think it would be better to just ship this plugins feature with all 3 references supported or add it in later?

I think we should remove support for branches and add it later ensuring it's properly supported. Otherwise, developers will start using it, get unexpected results, and report it as a bug. I'd rather wait and implement it handling that scenario gracefully What do you think?

I think we should focus on getting plug-ins working, release them to the wild, and see how users use them. Based on the feedback,  we can iterate on the implementation and add features like support for branches.

pepibumur commented:
By the way, you'll have to rebase from the main branch to fix the failing GitHub workflow associated to the Next website. The failing unit test seems legit though :)

luispadron commented:
@pepibumur, That sounds good. Ill remove the branch support for now and see whats up with CI when I get a chance later today.
